### PR TITLE
[ML] Improvements to sparse count modelling

### LINKED ISSUE #696: [ML] Improve modeling of periodic, sparse data
Prospective customer use case has once-daily sampled data of a program executed at a certain time. 

image

Tried to model with 4h bucket span in order to get somewhat timely alerting if a gap exists (a monitored program does not run), but the modeling isn't adequate. Anomalies are not raised at the time of the gap of execution:

image

The modeling and the anomaly results are better with a 1d bucket_span:

image

But, the downside is an increased "delay to alert" in this case.

### PR DESCRIPTION
**Pull request by tveasey**
This simplifies sparse count and sum modelling and migrates to always updating the time series model, but using a weight which decreases in proportion to the number of empty buckets. This means we simply smoothly transition to modelling non-empty buckets for sparse data.

I've also removed the correction to the probability which accounts for the fraction of non-empty buckets: we have the rare function anyway if this is the primary concern. Finally, I changed periodicity testing so that it approximates the old behaviour, i.e. it tends to ignoring empty buckets as their proportion increases.

Closes #696.

### CODE REVIEW

File: lib/maths/CPeriodicityHypothesisTests.cc [hunk 1]
```diff
@@ -1839,10 +1838,12 @@ bool CPeriodicityHypothesisTests::testPeriodWithScaling(const TTimeTimePr2Vec& w
     }
 
     // Compute the degrees of freedom given the alternative hypothesis.
-    TSizeVec repeats(calculateRepeats(windows, period_, m_BucketLength, values));
-    double b{static_cast<double>(
-        std::count_if(repeats.begin(), repeats.end(),
-                      [](std::size_t repeat) { return repeat > 0; }))};
+    double b{[&] {
```
Comment (edsavage) [inline lib/maths/CPeriodicityHypothesisTests.cc]: Can we avoid using the default capture mode here?

File: lib/maths/CPeriodicityHypothesisTests.cc [hunk 2]
```diff
@@ -2056,13 +2057,9 @@ bool CPeriodicityHypothesisTests::testPartition(const TTimeTimePr2Vec& partition
...
 
     auto calculateMeanRepeats = [&](const TTimeTimePr2Vec& w, core_t::TTime p) {
-        TSizeVec repeats{calculateRepeats(w, p, m_BucketLength, values)};
-        return CBasicStatistics::mean(
-            std::accumulate(repeats.begin(), repeats.end(), TMeanAccumulator{},
-                            [](TMeanAccumulator mean, std::size_t r) {
-                                mean.add(static_cast<double>(r));
-                                return mean;
-                            }));
+        TMeanAccumulator result;
```
Comment (edsavage) [inline lib/maths/CPeriodicityHypothesisTests.cc]: ... and here (I realise it's not part of this change)

File: lib/maths/CPeriodicityHypothesisTests.cc [hunk 3]
```diff
@@ -2143,13 +2140,11 @@ bool CPeriodicityHypothesisTests::testVariance(const TTimeTimePr2Vec& window,
...
     LOG_TRACE(<< "  autocorrelationThreshold = " << stats.s_AutocorrelationThreshold);
 
-    TSizeVec repeats{calculateRepeats(window, period_, m_BucketLength, buckets)};
-    meanRepeats = CBasicStatistics::mean(
-        std::accumulate(repeats.begin(), repeats.end(), TMeanAccumulator{},
-                        [](TMeanAccumulator mean, std::size_t repeat) {
-                            mean.add(static_cast<double>(repeat));
-                            return mean;
-                        }));
+    meanRepeats = [&] {
```
Comment (edsavage) [inline lib/maths/CPeriodicityHypothesisTests.cc]: ditto

File: lib/maths/CTimeSeriesModel.cc [hunk 4]
```diff
@@ -2619,18 +2549,13 @@ bool CMultivariateTimeSeriesModel::probability(const CModelProbabilityParams& pa
     };
     using TJointProbability2Vec = core::CSmallVector<SJointProbability, 2>;
 
-    auto updateJointProbabilities = [&](maths_t::EProbabilityCalculation calculation,
-                                        double value_, const TDouble10Vec2Vec& pl,
+    auto updateJointProbabilities = [&](const TDouble10Vec2Vec& pl,
```
Comment (edsavage) [inline lib/maths/CTimeSeriesModel.cc]: ditto

File: lib/model/CProbabilityAndInfluenceCalculator.cc [hunk 5]
```diff
@@ -146,43 +146,33 @@ double ratio(double numerator, double denominator, double zeroDividedByZero) {
...
 
 // Functions to compute influence based on different criteria
+
 //! \brief Computes the value of summed statistics on the set difference.
 class CValueDifference {
 public:
     //! Features.
     bool operator()(const TDouble2Vec& v,
-                    double n,
+                    double /*n*/,
```
Comment (edsavage) [inline lib/model/CProbabilityAndInfluenceCalculator.cc]: Is it possible to remove the commented out parameters entirely?

File: lib/maths/CPeriodicityHypothesisTests.cc (hunk 1 above)
Comment (tveasey) [reply lib/maths/CPeriodicityHypothesisTests.cc]: Done in [5d3c3a3](https://github.com/elastic/ml-cpp/pull/721/commits/5d3c3a3b79319afb4949e5ad10666d9adf1036d3).

File: lib/maths/CPeriodicityHypothesisTests.cc (hunk 3 above)
Comment (tveasey) [reply lib/maths/CPeriodicityHypothesisTests.cc]: Done in [5d3c3a3](https://github.com/elastic/ml-cpp/pull/721/commits/5d3c3a3b79319afb4949e5ad10666d9adf1036d3).

File: lib/maths/CTimeSeriesModel.cc (hunk 4 above)
Comment (tveasey) [reply lib/maths/CTimeSeriesModel.cc]: Done in [5d3c3a3](https://github.com/elastic/ml-cpp/pull/721/commits/5d3c3a3b79319afb4949e5ad10666d9adf1036d3).
